### PR TITLE
acceptable casks: don't allow beta versions

### PR DESCRIPTION
**Pull request by SMillerDev**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This brings the rules in line with Homebrew/core and makes them easier to understand/enforce since there are already audits that warn against this.

### CODE REVIEW

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by vitorgalvao** [comment]
Disagree. This rule has worked well from its inception, and I see no strong reason to change it but strong reasons not to.

It’s not uncommon for apps to be stable yet be called betas. It’s also not uncommon for such apps to become very popular very fast (e.g. something new that explodes on Hacker News or Product Hunt). And there have been more than a handful of situations where upstream officially recommends an “unstable” version because in practice it’s more stable, or fixes a major bug, or is the one which supports the latest (and stable) macOS release.

As far as I’m aware the audits for this are limited—they may work well for formula, but most casks aren’t well-organised GitHub repos with machine-readable definitions of “beta”.

This rule change would create a dance where we need to add the cask in the versions repo under one name (`*-beta`), then move it to the main repo at a later point with a different name. Then do it a few more time in the cases mentioned above. *And we don’t have support for `cask_renames.json`*, so we would be breaking casks even more and placing a bigger burden on the user.

I see a lot of downside, and no upside. This is one of those rare situations where consistency across formulae and casks would make Homebrew less convenient and predictable, not more.

**Comment by SMillerDev** [comment]
> It’s not uncommon for apps to be stable yet be called betas.

You might see them as stable, but by definition this is not true.

> It’s also not uncommon for such apps to become very popular very fast (e.g. something new that explodes on Hacker News or Product Hunt). And there have been more than a handful of situations where upstream officially recommends an “unstable” version because in practice it’s more stable, or fixes a major bug, or is the one which supports the latest (and stable) macOS release.

We can still allow them as an exception if we think they can be stable soon.

> As far as I’m aware the audits for this are limited—they may work well for formula, but most casks aren’t well-organised GitHub repos with machine-readable definitions of “beta”.

For most of the beta submissions in the past couple of weeks they actually have been this as far as I saw.

> This rule change would create a dance where we need to add the cask in the versions repo under one name (*-beta), then move it to the main repo at a later point with a different name. Then do it a few more time in the cases mentioned above. 

> And we don’t have support for cask_renames.json, so we would be breaking casks even more and placing a bigger burden on the user.

Both of these are only a burden if the user wants to move from the beta stream to stable.

> I see a lot of downside, and no upside.

The main upside is that the current ruleset for homebrew cask is full of inconsistencies that make it harder to help out for maintainers that haven't written those rules. And it's much harder to have CI help in maintenance, since audits for rules that apply only half of the time are hard to write.

**Comment by vitorgalvao** [comment]
> You might see them as stable, but by definition this is not true.

No, I’m talking about cases where the developers specifically recommend those betas over the stable version, but for one reason or another have tagged it as a beta.

> We can still allow them as an exception if we think they can be stable soon.

Which is every case this rule is targeting! That is precisely why the rule exists. What you’re suggesting ends up being changing the letter of the law but keeping everything the same in practice.

> For most of the beta submissions in the past couple of weeks they actually have been this as far as I saw.

I believe you. That doesn’t mean that’s the trend in the long run (over years of observation, it doesn’t seem to be, but I’d like to see hard data), or that it will continue.

> Both of these are only a burden if the user wants to move from the beta stream to stable.

Which is (close to) **everyone** in these cases! In the vast majority of these situations the beta stops existing (or stops being developed) once the app becomes stable. So a user who want the latest stable version of an app (i.e. everyone using the main cask tap) but is willing to use the beta until then (i.e. the closest thing to a stable at the time) would now need to keep up with the app’s versioning upstream and know when to switch casks. That is an awful and inconvenient experience.

> The main upside is that the current ruleset for homebrew cask is full of inconsistencies that make it harder to help out for maintainers that haven't written those rules. And it's much harder to have CI help in maintenance, since audits for rules that apply only half of the time are hard to write.

Except the rule in question is neither inconsistent nor hard to grasp. “Is there no stable version? Allow it in the stable repo”.

With the necessary added dance this change proposes, you’re not reducing the burden to maintainers, you’re increasing it! Because now we need to keep track of multiple casks over multiple repos, move one, delete the other, and deal with the confused users.

This rule was never a source of user confusion. Users tend to submit to the main repo anyway, so it’s *less* work for maintainers to allow it at submission because it will sort itself out in time when a stable version is released.

File: docs/Acceptable-Casks.md
Comment (vitorgalvao) [inline docs/Acceptable-Casks.md:31]: This should make the rule clearer:

```suggestion
When software is only available as a beta, development, or unstable version, its cask can go in the main repo. After a stable version is out, further updates to unstable versions will no longer be accepted.
```

We should keep the bit about exceptions further down. E.g. if there’s a major security hole in a stable version that is fixed in a beta.

**Comment by SMillerDev** [comment]
> No, I’m talking about cases where the developers specifically recommend those betas over the stable version, but for one reason or another have tagged it as a beta.

The recommended version doesn't have to be stable.

> Which is every case this rule is targeting! That is precisely why the rule exists. What you’re suggesting ends up being changing the letter of the law but keeping everything the same in practice.

But the current rule is: "everything is allowed if there is no stable version yet". The new implementation would be: "maintainers can allow beta versions if they choose to". In the latter there is no way for users to demand inclusion "because the rule says so".

> I believe you. That doesn’t mean that’s the trend in the long run (over years of observation, it doesn’t seem to be, but I’d like to see hard data), or that it will continue.

That's fair.

> Which is (close to) everyone in these cases! In the vast majority of these situations the beta stops existing (or stops being developed) once the app becomes stable. So a user who want the latest stable version of an app (i.e. everyone using the main cask tap) but is willing to use the beta until then (i.e. the closest thing to a stable at the time) would now need to keep up with the app’s versioning upstream and know when to switch casks. That is an awful and inconvenient experience.

We can use a `deprecated`/`discontinued` for caveat for this. That way users explicitly have to switch release channels.

> Except the rule in question is neither inconsistent nor hard to grasp. “Is there no stable version? Allow it in the stable repo”.

If upstream suggests installing a beta over stable, as you pointed out above, this is not covered by the rule as it is written. The rule as it is currently worded is only a stick for users to get their beta software into the main cask channel. I feel that leaving it explicitly up to maintainers to make exceptions is a better process.

> With the necessary added dance this change proposes, you’re not reducing the burden to maintainers, you’re increasing it! Because now we need to keep track of multiple casks over multiple repos, move one, delete the other, and deal with the confused users.

It's not a maintainers task to update/add/delete software. If users want software to be available it's up to them to submit a stable version.

> This rule was never a source of user confusion. Users tend to submit to the main repo anyway, so it’s less work for maintainers to allow it at submission because it will sort itself out in time when a stable version is released.

I think users tend to submit to the main repo because there is a rule telling them to.

File: docs/Acceptable-Casks.md
Comment (SMillerDev) [reply docs/Acceptable-Casks.md:31]: But this goes against your argument of "often upstream recommends a beta"

**Comment by vitorgalvao** [comment]
> But the current rule is: "everything is allowed if there is no stable version yet".

Yes, and my argument is that this has always been positive. I don’t recall a single instance of regretting this rule.

> In the latter there is no way for users to demand inclusion "because the rule says so".

Is that issue? Again, I don’t recall an instance where this rule was a problem. Either way, your “Sometimes there are exceptions” point at the end covers that:

User: “You need to accept it because the rule says so”.

Maintainer: “The rule also says we may reject the cask even if it follows the rules”.

> We can use a `deprecated`/`discontinued` for caveat for this. That way users explicitly have to switch release channels.

Which, again, is a bad experience for users and more work for us.

> If upstream suggests installing a beta over stable, as you pointed out above, this is not covered by the rule as it is written.

It is. It’s the `or in cases where such a version is the general standard` part.

> The rule as it is currently worded is only a stick for users to get their beta software into the main cask channel.

Again, is that a problem? I don’t recall that ever being an issue. And again, your “Sometimes there are exceptions” point covers that.

> It's not a maintainers task to update/add/delete software. If users want software to be available it's up to them to submit a stable version.

It’s our task to review those changes, though. And the harder the dance users have to go through to make the change, the more work for us. Especially when that change has to be coordinated over two repos.

> I think users tend to submit to the main repo because there is a rule telling them to.

They tend to submit to the main repo because they haven’t read the documentation and don’t know the other repo exists.

Right now, a user submitting an unstable version to the main repo is only a problem if the app has a stable version. That makes those errors relatively rare. With the rule change, that error would become more common (again, more work for us).

**Comment by MikeMcQuaid** [comment]
> It’s not uncommon for apps to be stable yet be called betas. It’s also not uncommon for such apps to become very popular very fast (e.g. something new that explodes on Hacker News or Product Hunt). And there have been more than a handful of situations where upstream officially recommends an “unstable” version because in practice it’s more stable, or fixes a major bug, or is the one which supports the latest (and stable) macOS release.

This applies equally to formulae. I think there's an inherent benefit in being consistent here.

> The new implementation would be: "maintainers can allow beta versions if they choose to". In the latter there is no way for users to demand inclusion "because the rule says so".

👍🏻 

> Right now, a user submitting an unstable version to the main repo is only a problem if the app has a stable version. That makes those errors relatively rare. With the rule change, that error would become more common (again, more work for us).

Then again: "this rule makes our life harder" is a legit argument so I don't have a strong feeling either way here.

**Comment by vitorgalvao** [comment]
> This applies equally to formulae. I think there's an inherent benefit in being consistent here.

That’s [addressed by my suggestion](https://github.com/Homebrew/brew/pull/11499#discussion_r646695986); which removes those cases from the rule.

> > The new implementation would be: "maintainers can allow beta versions if they choose to". In the latter there is no way for users to demand inclusion "because the rule says so".
> 
> 👍🏻

Again, *that has never been an issue!* A hypothetical which has never happened in a decade isn’t sound reasoning for changing a rule.

> Then again: "this rule makes our life harder" is a legit argument

And again I ask, *does it*? Because from the myriad of maintainers that passed through HBC, this is the first time this has ever come up. Do the other HB maintainers who are helping in HBC agree that this rule has caused them problems? And if so, in what way does [the suggested edit](https://github.com/Homebrew/brew/pull/11499#discussion_r646695986) not fix that concern?

This rule makes our life *easier*; *changing it* is what would make it harder, for the all the reasons I’ve listed above. Right now we have a well-defined logical case that conforms to user expectations. The suggestion is to replace it with a coordinated dance over two repos, breaking casks and putting more work on users *and us*. It’s worse for everyone.

**Comment by SMillerDev** [comment]
> Again, that has never been an issue! A hypothetical which has never happened in a decade isn’t sound reasoning for changing a rule.

I've already seen a PRs with a quote of that rule as justification why it should be added. I don't think it's that hypothetical

https://github.com/Homebrew/homebrew-cask/pull/106279#pullrequestreview-677338340

**Comment by MikeMcQuaid** [comment]
> _Assume kindness_. Why are you starting from the assumption that the user would be demanding or unreasonable in any way? There is zero reason to think that would be the case. All they did was be proactive, _which is something we want_.

Similarly: when @SMillerDev didn't use the words "demanding" or "unreasonable": why assume that is his intent/thought process?

**Comment by vitorgalvao** [comment]
> Similarly: when @SMillerDev didn't use the words "demanding" or "unreasonable": why assume that is his intent/thought process?

https://github.com/Homebrew/brew/pull/11499#issuecomment-856033163:

> In the latter there is no way for users to demand inclusion "because the rule says so".

**Comment by MikeMcQuaid** [comment]
@vitorgalvao You're being a bit pedantic here so I'll be more explicit: your language is escalating things here in a way that's not necessary. Let's please focus on what the document says or doesn't say rather than what @SMillerDev thinks or doesn't think and about whether people are or are not "pricks".

**Comment by vitorgalvao** [comment]
@MikeMcQuaid I appreciate the sentiment. I don’t think there’s anything to gain from continuing that tangent here.

There have been a couple of other maintainer reactions on this thread, so I propose we leave the final word to someone else who hasn’t commented yet.

**Comment by SMillerDev** [comment]
While I still think this would help ease people who are used to the core tap into cask contributions, remedying the reported lack of cask maintainers, if the majority disagrees I'll just close it.

**Pull request closed without merging**
